codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerHttpSensitivityGenerator.kt

```diff
@@ -0,0 +1,202 @@
+            }
+
+            rustBlockTemplate("#{SmithyHttpServer}::logging::HeaderMarker", *codegenScope) {
+                rust("value, key_suffix")
+            }
+        }
+    }
+
+    internal fun renderQueryClosure(writer: RustWriter, queries: List<HttpQueryTrait>) {
+        writer.withBlockTemplate("|name: &str| #{SmithyHttpServer}::logging::QueryMarker { key: false, value: matches!(name,", ") }", *codegenScope) {
+            val matches = queries.map { it.value }.distinct().map { "\"$it\"" }.joinToString("|")
+            rust(matches)
+        }
+    }
+
+    internal fun renderQueryParamsClosure(writer: RustWriter) {
+        writer.rustBlock("|_: &str|") {
+            rustTemplate(
+                "#{SmithyHttpServer}::logging::QueryMarker { key: true, value: true }", *codegenScope
+            )
+        }
+    }
+
+    internal fun renderPathClosure(writer: RustWriter, indexes: List<Int>) {
+        writer.rustBlock("|index: usize|") {
+            withBlock("matches!(index,", ")") {
+                val matches = indexes.map { "$it" }.joinToString("|")
+                rust(matches)
+            }
+        }
+    }
+
+    // Find member shapes which are sensitive and enjoy a trait `T`.
+    internal inline fun <reified T : Trait> findSensitiveBound(rootShape: Shape): List<MemberShape> {
+        return Walker(model)
+            .walkShapes(rootShape, {
+                // Do not traverse upwards or beyond a sensitive trait
+                it.getDirection() == RelationshipDirection.DIRECTED && it.getShape().getTrait<SensitiveTrait>() == null
+            })
+            .filter {
+                it.getTrait<SensitiveTrait>() != null
+            }
+            .flatMap {
+                Walker(model)
+                    .walkShapes(it, {
+                        // Do not traverse upwards
+                        it.getDirection() == RelationshipDirection.DIRECTED
+                    })
+                    .filter {
+                        it.getTrait<T>() != null
+                    }
+                    .map {
+                        it as? MemberShape
+                    }
+                    .filterNotNull()
+            }
+    }
+
+    // Find traits (applied to member shapes) which are sensitive.
+    internal inline fun <reified T : Trait> findSensitiveBoundTrait(rootShape: Shape): List<T> {
+        return findSensitiveBound<T>(rootShape).map { it.getTrait<T>() }.filterNotNull()
+    }
+
+    internal fun findUriLabelIndexes(httpTrait: HttpTrait, inputShape: Shape): List<Int> {
+        val uriLabels: Map<String, Int> = httpTrait
+            .uri
+            .getSegments()
+            .withIndex()
+            .filter { (_, segment) -> segment.isLabel() }
+            .map { (index, segment) -> Pair(segment.getContent(), index) }
+            .toMap()
+        return findSensitiveBound<HttpLabelTrait>(inputShape)
+            .map { uriLabels.get(it.getMemberName()) }
+            .filterNotNull()
```

Comment:
That's right. In rust I would `.expect("cannot fail as we checked trait previously")`. In Kotlin should I `!!`?

Reply:
In Kotlin, the following is a much nicer way to do this than `!!` since you can add a reason that becomes part of the error message:
```kotlin
checkNotNull(thingToCheck) { "thingToCheck shouldn't be null for <reason>" }
```
See [`checkNotNull`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/check-not-null.html).